src/util/metadata/mod.rs

```diff
@@ -225,3 +225,17 @@ pub mod side_metadata;
 pub(crate) mod log_bit;
 
 pub use global::*;
+
+// We currently only support metadata that has a length equal to or smaller than the architecture's word size.
+// That means we do not support metadata of more than 8 bytes, and we do not support side metadata of 8 bytes on 32 bits architectures.
+// This macro is used to check word size. It makes sure that the given block is only executed on 64 bits architectures, and panics on other architectures.
+macro_rules! only_available_on_64bits {
+    ($b: block) => {
+        if cfg!(target_pointer_width = "64")
```

Comment:
Actually `AtomicU64` may be available even if the pointer width is 32 bits.  For example, on ARMv7, the `LDRD` and `STRD` instructions can load/store a 64-bit aligned region atomically.  `LDREXD` paired with `STREXD` can implement atomic update (including compare-exchange( at 64-bit granularity.

I suggest we do not add `only_available_on_64bits` now.  We assume `AtomicU64` is always available.  By doing this, when we ever compile mmtk-core on a target that doesn't support `AtomicU64`, we will get a compilation error, and we can fix that later.

A more precise feature testing macro is `target_has_atomic_load_store` for load and store, and `target_has_atomic` for atomic read-modify-write operations.  At this moment, both of them are unstable features, but may be stablized soon.

Reply:
Yeah. As Kunshan and I discussed, we will make 64bits metadata usable on 32bits architecture where it is possible. This requires us changing the metadata-related API (we cannot use `usize` as arguments and return values if we want to use 64bits values on 32bits architectures)

Reply:
I did some refactoring on this. Basically I will introduce a new trait `MetadataValue`: when storing and loading from metadata, this type is used to describe the value. `u8, u16, u32, u64` implement this trait.

The old code looks like this (will be referred as *untyped*):
https://github.com/mmtk/mmtk-core/blob/d8bcf90991224cfe06153689b14c3e1eea0e025e/src/util/metadata/side_metadata/global.rs#L671-L707

This gets compiled to (in a use case of 1-bit alloc bit):
```asm
mmtk::util::alloc_bit::is_alloced_object_unsafe_untyped:
 mov     rcx, rdi
 mov     rax, rdi
 shr     rax, 6
 shr     cl, 3
 and     cl, 7
 movabs  rdx, 6597069766656
 movzx   eax, byte, ptr, [rax, +, rdx]
 mov     edx, 1
 shl     edx, cl
 and     edx, eax
 shr     rdx, cl
 cmp     edx, 1
 sete    al
 ret
```

After introducing the `MetadataValue` type, and extracting some common code, the new code looks like this (will be referred as *typed*):
```rust
#[inline(always)]
pub unsafe fn typed_load<T: MetadataValue>(metadata_spec: &SideMetadataSpec, data_addr: Address) -> T {
    side_metadata_access::<T, _, _, _>(metadata_spec, data_addr, || {
        let meta_addr = address_to_meta_address(metadata_spec, data_addr);
        let bits_num_log = metadata_spec.log_num_of_bits;
        if bits_num_log < 3 {
            let lshift = meta_byte_lshift(metadata_spec, data_addr);
            let mask = meta_byte_mask(metadata_spec) << lshift;
            let byte_val = meta_addr.load::<u8>();

            num_traits::FromPrimitive::from_u8((byte_val & mask) >> lshift).unwrap()
        } else {
            meta_addr.load::<T>()
        }
    }, |v| {
        #[cfg(feature = "extreme_assertions")]
        sanity::typed_verify_load(metadata_spec, data_addr, v);
    })
}

fn assert_int_type<T: Bits>(spec: &SideMetadataSpec) {
    let log_b = spec.log_num_of_bits;
    match log_b {
        _ if log_b < 3 => assert_eq!(T::LOG2, 3),
        3..=6 => assert_eq!(T::LOG2, log_b as u32),
        _ => unreachable!("side metadata > {}-bits is not supported", 1 << log_b),
    }
}

#[inline(always)]
fn side_metadata_access<T: MetadataValue, R: Copy, F: Fn() -> R, V: Fn(R)>(spec: &SideMetadataSpec, data_addr: Address, access_func: F, verify_func: V) -> R {
    #[cfg(feature = "extreme_assertions")]
    let _lock = sanity::SANITY_LOCK.lock().unwrap();

    if cfg!(debug_assertions) {
        assert_int_type::<T>(spec);
        ensure_metadata_is_mapped(spec, data_addr);
    }

    let ret = access_func();

    verify_func(ret);

    return ret;
}
```

This gets compiled to (in the same use case as above):
```asm
mmtk::util::alloc_bit::is_alloced_object_unsafe_typed:
 mov     rcx, rdi
 mov     rax, rdi
 shr     cl, 3
 and     cl, 7
 mov     dl, 1
 shl     dl, cl
 shr     rax, 6
 movabs  rsi, 6597069766656
 and     dl, byte, ptr, [rax, +, rsi]
 shr     dl, cl
 cmp     dl, 1
 sete    al
 ret
```

I think performance-wise, the new code should be mostly the same as before. And the new code should be much clearer and has less duplicated code. I plan to finish the refactoring in this PR.